### ISSUE TITLE
MSSQL varchar(MAX) support

### ISSUE THREAD

**Issue by gbkwiatt** [knex/knex#3871]
Is there any way to make schema and migration to create varchar(MAX) ?

currently it's max 8000 and does not allow string value (it defaults to 255).

**Comment by gbkwiatt** [comment]
I did some change in columncompiler.js in dialects/mssql/schema

```javascript
  varchar(length) {
    if(length === 'MAX') return 'varchar(MAX)';   //added that 
    return `varchar(${this._num(length, 255)})`;
  },
```

**Comment by kibertoad** [comment]
Can you submit pr with test for this change?

**Comment by gbkwiatt** [comment]
well actually it's my bad-ish. 
I prefer and suggest varchar in mssql not nvarchar, so I've changed those. 
And then text should be varchar(MAX)

or at least there should be a way to force knex using varchar instead of nvarchar.

**Comment by kibertoad** [comment]
I agree. PR is welcome.

**Comment by MertcanYuzer** [comment]
https://github.com/knex/knex/issues/2533#issuecomment-1341640251
For those who still need to use NVARCHAR(MAX) type, you can try table.text("column_name"). I think it is an ideal solution.

**Comment by RachelAmbler** [comment]
> [#2533 (comment)](https://github.com/knex/knex/issues/2533#issuecomment-1341640251)
> For those who still need to use NVARCHAR(MAX) type, you can try table.text("column_name"). I think it is an ideal solution.

No, it's not, it's a fudge. MS marked `Text` and `NText` as deprecated in 2005. It's getting close to 20 years later and there's zero reason why Knex should still be using those datatypes.

**Comment by mercmobily** [comment]
Hi there! Thanks for raising this and for your patience.

On MSSQL, Knex maps string columns through the dialect column compiler. The current implementation only supports numeric lengths for varchar‑like columns and defaults to 255 when length is omitted, so there is no path that emits nvarchar(max)/varchar(max) from the standard schema builder API. Specifically, `varchar(length)` always emits `nvarchar(<number>)` and does not recognise a sentinel for MAX, and omitting the length chooses 255 rather than MAX. See lib/dialects/mssql/schema/mssql-columncompiler.js:23 and lib/dialects/mssql/schema/mssql-columncompiler.js:27–33 for the current behaviour, and lib/dialects/mssql/schema/mssql-columncompiler.js:167–174 for related MAX mappings used for text/json types.

In the code: lib/dialects/mssql/schema/mssql-columncompiler.js:23 defines `varchar(length)` and returns `nvarchar(${toNumber(length, 255)})`. There is no condition to return `(n)varchar(max)`. lib/dialects/mssql/schema/mssql-columncompiler.js:170–174 shows MAX mappings only for `text`, `mediumtext`, `longtext`, and `json/jsonb` → `nvarchar(max)`. In contrast, binary with omitted length becomes `varbinary(max)` (lib/dialects/mssql/schema/mssql-columncompiler.js:40–44), underscoring that MAX handling is explicit per‑type.

There are no CLOSED PRs in‑tree adding `(n)varchar(max)` emission via the schema builder. Users commonly use either `specificType('nvarchar(max)')` (or `'varchar(max)'` if Unicode is not desired) or use `text()` where applicable. No NG_ master issue was found for this exact feature; the closest related items are MSSQL type feature requests without MAX handling specifics.

The request is valid; Knex does not currently expose a builder path for `(n)varchar(max)`. Recommended immediate workarounds are:
- Use `table.specificType('col', 'nvarchar(max)')` or `specificType('varchar(max)')` for non‑Unicode.
- Use `table.text('col')` where semantics allow, which already maps to `nvarchar(max)` on MSSQL.

A minimal, backwards‑compatible enhancement would be to treat `length: null | 'max'` as MAX for MSSQL in `varchar()` and to add `nvarchar(length)` explicitly.
